File: src/backend/app/domain/services/dashboard_widget_catalog.py
"""REQ-045 — server-authoritative dashboard widget catalog.

Holds the backend-side widget metadata (category, experience-level default,
sizing, gating module) and resolves per-user *availability* from the gates
that only the server knows about:

- REQ-042 module visibility (a widget whose ``required_module`` the user hid
  is unavailable),
- REQ-027 Light-Mode (``community_activity`` is filtered; ``daily_tip`` needs
  a whitelisted AI provider),
- REQ-024 permissions (dashboard widgets are read-only; every tenant member
  may read them, so no extra role gate is applied here today).

The widget-key set MUST stay in sync with the frontend
``dashboardWidgetCatalog`` and with ``KNOWN_WIDGET_KEYS`` in
``user_preference_service`` (contract test, REQ-045 §6).
"""

from __future__ import annotations

from dataclasses import dataclass, field

from app.common.enums import ExperienceLevel, ModuleVisibilityState

# i18n keys the frontend resolves for the greyed-out "why not available" hint.
REASON_MODULE_HIDDEN = "dashboard.gate.moduleHidden"
REASON_LIGHT_MODE = "dashboard.gate.lightMode"
REASON_AI_DISABLED = "dashboard.gate.aiDisabled"


@dataclass(frozen=True)
class WidgetMeta:
    """Static, mode-independent metadata for one dashboard widget."""

    widget_key: str
    category: str
    default_level: ExperienceLevel
    default_size: dict[str, int]
    min_size: dict[str, int]
    max_size: dict[str, int]
    required_module: str | None = None
    light_gated: bool = False  # REQ-027: not selectable in Light mode at all
    requires_ai: bool = False  # REQ-031: needs a whitelisted AI provider


def _sz(w: int, h: int) -> dict[str, int]:
    return {"w": w, "h": h}


# Order matters only for a stable catalog response; grouping is by ``category``.
WIDGET_CATALOG: tuple[WidgetMeta, ...] = (
    WidgetMeta("quick_actions", "essentials", ExperienceLevel.BEGINNER, _sz(12, 2), _sz(4, 2), _sz(12, 4)),
    WidgetMeta("tasks_today", "essentials", ExperienceLevel.BEGINNER, _sz(4, 4), _sz(2, 3), _sz(8, 8), "tasks"),
    WidgetMeta("care_reminders", "essentials", ExperienceLevel.BEGINNER, _sz(4, 4), _sz(2, 3), _sz(8, 8), "care"),
    WidgetMeta(
        "active_plants_summary", "essentials", ExperienceLevel.BEGINNER, _sz(4, 3), _sz(2, 2), _sz(8, 6), "plants"
    ),
    WidgetMeta("onboarding_progress", "essentials", ExperienceLevel.BEGINNER, _sz(4, 3), _sz(2, 2), _sz(12, 4)),
    WidgetMeta(
        "daily_tip", "insights", ExperienceLevel.BEGINNER, _sz(4, 4), _sz(2, 3), _sz(8, 8), "ai", requires_ai=True
    ),
    WidgetMeta("weather_forecast", "insights", ExperienceLevel.BEGINNER, _sz(4, 3), _sz(2, 2), _sz(8, 6)),
    WidgetMeta("winter_protection", "cultivation", ExperienceLevel.BEGINNER, _sz(6, 4), _sz(3, 3), _sz(12, 8), "care"),
    WidgetMeta("ipm_alerts", "cultivation", ExperienceLevel.INTERMEDIATE, _sz(4, 4), _sz(2, 3), _sz(8, 8), "ipm"),
    WidgetMeta(
        "harvest_forecast", "insights", ExperienceLevel.INTERMEDIATE, _sz(4, 4), _sz(2, 3), _sz(8, 8), "harvest"
    ),
    WidgetMeta(
        "next_calendar_events",
        "cultivation",
        ExperienceLevel.INTERMEDIATE,
        _sz(4, 4),
        _sz(2, 3),
        _sz(8, 8),
        "calendar",
    ),
    WidgetMeta(
        "community_activity",
        "insights",
        ExperienceLevel.INTERMEDIATE,
        _sz(4, 4),
        _sz(2, 3),
        _sz(8, 8),
        light_gated=True,
    ),
    WidgetMeta("sensor_live", "monitoring", ExperienceLevel.EXPERT, _sz(4, 4), _sz(2, 3), _sz(8, 8), "sensors"),
    WidgetMeta("tank_status", "monitoring", ExperienceLevel.EXPERT, _sz(4, 4), _sz(2, 3), _sz(8, 8), "tanks"),
    WidgetMeta("phase_timeline", "cultivation", ExperienceLevel.EXPERT, _sz(8, 4), _sz(3, 3), _sz(12, 8), "plants"),
    WidgetMeta("vpd_gauge", "monitoring", ExperienceLevel.EXPERT, _sz(3, 4), _sz(2, 3), _sz(6, 8), "sensors"),
    WidgetMeta("plant_grid", "cultivation", ExperienceLevel.EXPERT, _sz(8, 5), _sz(4, 3), _sz(12, 10), "plants"),
)

WIDGET_BY_KEY: dict[str, WidgetMeta] = {w.widget_key: w for w in WIDGET_CATALOG}


@dataclass(frozen=True)
class WidgetAvailability:
    """Resolved catalog entry for one user (metadata + availability)."""

    meta: WidgetMeta
    available: bool
    unavailable_reason: str | None = field(default=None)


def resolve_widget_catalog(
    *,
    mode: str,
    module_visibility: dict[str, ModuleVisibilityState],
) -> list[WidgetAvailability]:
    """Resolve availability of every widget for the calling user.

    ``mode`` is ``settings.kamerplanter_mode`` ("light" | "full"). AI-gated
    widgets require a whitelisted provider, which today is only present in full
    mode (REQ-027 §2.1); Light mode therefore disables ``daily_tip`` and
    ``community_activity``.
    """

    is_light = mode == "light"
    resolved: list[WidgetAvailability] = []
    for meta in WIDGET_CATALOG:
        available = True
        reason: str | None = None
        if meta.required_module and module_visibility.get(meta.required_module) == ModuleVisibilityState.DISABLED:
            available, reason = False, REASON_MODULE_HIDDEN
        elif is_light and meta.light_gated:
            available, reason = False, REASON_LIGHT_MODE
        elif meta.requires_ai and is_light:
            available, reason = False, REASON_AI_DISABLED
        resolved.append(WidgetAvailability(meta=meta, available=available, unavailable_reason=reason))
    return resolved

File: src/frontend/src/config/dashboardWidgetCatalog.ts
import type { ExperienceLevel, DashboardLayout, WidgetPlacement } from '@/api/types';
import type { ModuleKey } from '@/config/moduleCatalog';

/**
 * REQ-045 — declarative dashboard widget catalog.
 *
 * Canonical frontend source of truth for the widgets a user can place on their
 * dashboard. Mirrors the backend registry (``KNOWN_WIDGET_KEYS`` /
 * ``dashboard_widget_catalog``); the shared key set is asserted by the contract
 * test (``dashboard-widgets.json``). Availability (module/permission/light-mode
 * gating) is server-authoritative and layered on top at runtime — this catalog
 * only holds the static, mode-independent metadata.
 */

export type WidgetKey =
  | 'quick_actions'
  | 'active_plants_summary'
  | 'tasks_today'
  | 'care_reminders'
  | 'daily_tip'
  | 'weather_forecast'
  | 'onboarding_progress'
  | 'winter_protection'
  | 'ipm_alerts'
  | 'harvest_forecast'
  | 'next_calendar_events'
  | 'community_activity'
  | 'tank_status'
  | 'phase_timeline'
  | 'plant_grid';

export interface WidgetSize {
  w: number;
  h: number;
}

export interface DashboardWidgetDefinition {
  key: WidgetKey;
  /** i18n key for the label, e.g. 'dashboard.widgets.tasks_today.label' */
  labelKey: string;
  /** i18n key for the explanatory description in the settings tab (UI-NFR-011) */
  descriptionKey: string;
  /** Grouping in the settings tab */
  category: string;
  /** Default set membership (REQ-021) */
  defaultLevel: ExperienceLevel;
  defaultSize: WidgetSize;
  minSize: WidgetSize;
  maxSize: WidgetSize;
  /** Gated via REQ-042; null = bound to core (always visible) */
  requiredModule: ModuleKey | null;
  /**
   * Issue #587 — widget surfaces sensor/actuator data (monitoring/telemetry) and
   * is additionally gated behind the per-user `smart_home_enabled` toggle. When
   * smart home is off the widget is neither rendered nor offered in the picker.
   */
  requiresSmartHome?: boolean;
  /** true → offers a per-widget config dialog */
  hasConfig: boolean;
  /**
   * Panel-level navigation target (issue #439). When set, the whole widget panel
   * becomes a link to this route in read-only mode (gated by module visibility,
   * REQ-042). null/undefined → the panel is non-interactive (read-only overview).
   */
  navigateTo?: string;
}

/** Grid columns per breakpoint (UI-NFR-001): Desktop / Tablet / Mobile. */
export const GRID_COLS_BY_BREAKPOINT = { lg: 12, md: 8, sm: 4 } as const;
export const DASHBOARD_BREAKPOINTS = ['lg', 'md', 'sm'] as const;
export const DASHBOARD_LAYOUT_SCHEMA_VERSION = 2;

function def(
  key: WidgetKey,
  category: string,
  defaultLevel: ExperienceLevel,
  defaultSize: WidgetSize,
  minSize: WidgetSize,
  maxSize: WidgetSize,
  requiredModule: ModuleKey | null,
  hasConfig = false,
  navigateTo?: string,
  requiresSmartHome = false,
): DashboardWidgetDefinition {
  return {
    key,
    labelKey: `dashboard.widgets.${key}.label`,
    descriptionKey: `dashboard.widgets.${key}.description`,
    category,
    defaultLevel,
    defaultSize,
    minSize,
    maxSize,
    requiredModule,
    requiresSmartHome,
    hasConfig,
    navigateTo,
  };
}

const s = (w: number, h: number): WidgetSize => ({ w, h });

export const dashboardWidgetCatalog: Record<WidgetKey, DashboardWidgetDefinition> = {
  // ── Essentials ──
  // h=4 (not 2): at 12 cols the tile Grid (xs:6/sm:4/md:2) still renders one
  // full row of 6 icon tiles + a title line; h=2 (2*44+16=104px) clipped that
  // content in both the edit grid (fixed ROW_HEIGHT) and, previously, the
  // read-only grid. h=4 matches maxSize.h, i.e. the widget can no longer be
  // grown further but starts at a height that actually fits its content.
  quick_actions: def('quick_actions', 'essentials', 'beginner', s(12, 4), s(4, 2), s(12, 4), null),
  tasks_today: def('tasks_today', 'essentials', 'beginner', s(4, 4), s(2, 3), s(8, 8), 'tasks', false, '/aufgaben/queue'),
  care_reminders: def('care_reminders', 'essentials', 'beginner', s(4, 4), s(2, 3), s(8, 8), 'tasks', false, '/aufgaben/queue'),
  active_plants_summary: def('active_plants_summary', 'essentials', 'beginner', s(4, 3), s(2, 2), s(8, 6), 'plants', false, '/pflanzen/plant-instances'),
  onboarding_progress: def('onboarding_progress', 'essentials', 'beginner', s(4, 3), s(2, 2), s(12, 4), null, false, '/onboarding'),

  // ── Insights ──
  daily_tip: def('daily_tip', 'insights', 'beginner', s(4, 4), s(2, 3), s(8, 8), 'ai'),
  weather_forecast: def('weather_forecast', 'insights', 'beginner', s(4, 3), s(2, 2), s(8, 6), null, true),
  harvest_forecast: def('harvest_forecast', 'insights', 'intermediate', s(4, 4), s(2, 3), s(8, 8), 'harvest', true, '/ernte/batches'),
  community_activity: def('community_activity', 'insights', 'intermediate', s(4, 4), s(2, 3), s(8, 8), null),

  // ── Cultivation ──
  winter_protection: def('winter_protection', 'cultivation', 'beginner', s(6, 4), s(3, 3), s(12, 8), 'overwintering', false, '/ueberwinterung/profile'),
  ipm_alerts: def('ipm_alerts', 'cultivation', 'intermediate', s(4, 4), s(2, 3), s(8, 8), 'ipm', false, '/pflanzenschutz/pests'),
  next_calendar_events: def('next_calendar_events', 'cultivation', 'intermediate', s(4, 4), s(2, 3), s(8, 8), 'calendar', false, '/kalender'),
  phase_timeline: def('phase_timeline', 'cultivation', 'expert', s(8, 4), s(3, 3), s(12, 8), 'plants', false, '/phasen/ablaeufe'),
  plant_grid: def('plant_grid', 'cultivation', 'expert', s(8, 5), s(4, 3), s(12, 10), 'plants', false, '/pflanzen/plant-instances'),

  // ── Monitoring ──
  // Issue #587: monitoring widgets surface sensor telemetry → smart-home-gated.
  tank_status: def('tank_status', 'monitoring', 'expert', s(4, 4), s(2, 3), s(8, 8), 'tanks', false, '/standorte/tanks', true),
};

/** Ordered list of categories for grouping in the settings tab. */
export const WIDGET_CATEGORIES: string[] = ['essentials', 'insights', 'cultivation', 'monitoring'];

/**
 * REQ-021 — default widget set per experience level. Higher levels are
 * additive (each includes all widgets of the levels below it).
 */
const BEGINNER_WIDGETS: WidgetKey[] = [
  'quick_actions',
  'tasks_today',
  'care_reminders',
  'active_plants_summary',
  'daily_tip',
  'winter_protection',
  'weather_forecast',
  'onboarding_progress',
];
const INTERMEDIATE_WIDGETS: WidgetKey[] = [
  ...BEGINNER_WIDGETS,
  'ipm_alerts',
  'harvest_forecast',
  'next_calendar_events',
  'community_activity',
];
const EXPERT_WIDGETS: WidgetKey[] = [
  ...INTERMEDIATE_WIDGETS,
  'tank_status',
  'phase_timeline',
  'plant_grid',
];

export const DEFAULT_WIDGETS_BY_LEVEL: Record<ExperienceLevel, WidgetKey[]> = {
  beginner: BEGINNER_WIDGETS,
  intermediate: INTERMEDIATE_WIDGETS,
  expert: EXPERT_WIDGETS,
};

let instanceCounter = 0;
/** Unique instance id for a *newly user-added* widget. NOT for the derived
 *  default layout — that must use deterministic ids (see resolveDefaultLayout)
 *  so re-deriving it across renders keeps stable React keys. */
export function newInstanceId(): string {
  instanceCounter += 1;
  return `w-${instanceCounter.toString(36)}-${Math.floor(performance.now()).toString(36)}`;
}

/**
 * Pack the given widget keys row-by-row into a 12-column ``lg`` grid using each
 * widget's ``defaultSize``. Produces a full layout (widgets + lg placements);
 * md/sm are left empty and derived client-side from lg by react-grid-layout.
 *
 * ``availableKeys`` (when provided) filters out server-unavailable widgets so a
 * fresh default never renders a gated widget.
 */
export function resolveDefaultLayout(level: ExperienceLevel, availableKeys?: ReadonlySet<string>): DashboardLayout {
  const cols = GRID_COLS_BY_BREAKPOINT.lg;
  const keys = DEFAULT_WIDGETS_BY_LEVEL[level].filter((k) => !availableKeys || availableKeys.has(k));

  // Deterministic ids: re-deriving the default across renders yields identical
  // React keys, so widgets reconcile instead of remounting (no dispatch loop).
  const widgets = keys.map((key) => ({ instance_id: `d-${key}`, widget_key: key, config: {} }));
  const placements: WidgetPlacement[] = [];
  let x = 0;
  let y = 0;
  let rowHeight = 0;
  for (let i = 0; i < widgets.length; i += 1) {
    const size = dashboardWidgetCatalog[keys[i]].defaultSize;
    const w = Math.min(size.w, cols);
    if (x + w > cols) {
      x = 0;
      y += rowHeight;
      rowHeight = 0;
    }
    placements.push({ instance_id: widgets[i].instance_id, x, y, w, h: size.h });
    x += w;
    rowHeight = Math.max(rowHeight, size.h);
  }

  return { schema_version: DASHBOARD_LAYOUT_SCHEMA_VERSION, widgets, placements: { lg: placements } };
}

/**
 * Migrate a possibly-v1 layout (schema_version 1 with widgets[].x/y/w/h and a
 * top-level ``columns``) to the v2 shape. Tolerant: unknown shapes fall back to
 * an empty layout. Idempotent for already-v2 layouts.
 */
export function migrateLayout(raw: unknown): DashboardLayout | null {
  if (!raw || typeof raw !== 'object') return null;
  const layout = raw as Record<string, unknown>;
  const widgets = Array.isArray(layout.widgets) ? (layout.widgets as Record<string, unknown>[]) : [];

  // Already v2: has a placements map.
  if (layout.placements && typeof layout.placements === 'object' && !Array.isArray(layout.placements)) {
    return {
      schema_version: DASHBOARD_LAYOUT_SCHEMA_VERSION,
      widgets: widgets.map((w, i) => ({
        instance_id: String(w.instance_id ?? `m-${w.widget_key}-${i}`),
        widget_key: String(w.widget_key),
        config: (w.config as Record<string, unknown>) ?? {},
      })),
      placements: layout.placements as DashboardLayout['placements'],
    };
  }

  // v1: positions live on the widgets themselves → read them as lg placements.
  const lg: WidgetPlacement[] = [];
  const migratedWidgets = widgets.map((w, i) => {
    // Deterministic fallback id (stable React keys across renders).
    const instanceId = String(w.instance_id ?? `m-${w.widget_key}-${i}`);
    if (typeof w.x === 'number' && typeof w.y === 'number') {
      lg.push({
        instance_id: instanceId,
        x: w.x as number,
        y: w.y as number,
        w: (w.w as number) ?? 4,
        h: (w.h as number) ?? 4,
      });
    }
    return {
      instance_id: instanceId,
      widget_key: String(w.widget_key),
      config: (w.config as Record<string, unknown>) ?? {},
    };
  });

  return {
    schema_version: DASHBOARD_LAYOUT_SCHEMA_VERSION,
    widgets: migratedWidgets,
    placements: lg.length ? { lg } : {},
  };
}

/**
 * Derive placements for a breakpoint. A breakpoint with its own placements
 * wins; otherwise fall back to lg (react-grid-layout auto-stacks narrower
 * breakpoints from lg). Widgets present in ``widgets`` but missing a placement
 * are appended in a single stacked column so nothing is ever dropped.
 */
export function placementsForBreakpoint(
  layout: DashboardLayout,
  breakpoint: 'lg' | 'md' | 'sm',
): WidgetPlacement[] {
  const own = layout.placements[breakpoint];
  const base = own && own.length ? own : (layout.placements.lg ?? []);
  const placed = new Set(base.map((p) => p.instance_id));
  const maxY = base.reduce((m, p) => Math.max(m, p.y + p.h), 0);
  const appended: WidgetPlacement[] = [];
  let y = maxY;
  for (const w of layout.widgets) {
    if (!placed.has(w.instance_id)) {
      appended.push({ instance_id: w.instance_id, x: 0, y, w: 4, h: 4 });
      y += 4;
    }
  }
  return [...base, ...appended];
}
